doc/developer/design/19042023_stash_migrations.md

```diff
@@ -0,0 +1,352 @@
+
+
+# Motivation
+[motivation]: #motivation
+
+Our overall goal is to create **"fearless Stash migrations"**. In other words, we want to make it
+so we can assign an issue that requires a Stash migration to a new-hire, and have confidence that
+if our builds and tests are passing, then the migration won't break anything in production.
+
+There have been several issues caused by Stash migrations:
+
+* [materialize#17824](https://github.com/MaterializeInc/materialize/issues/17824)
+  * General difficulty in writing Stash migrations, see [materialize#18719](https://github.com/MaterializeInc/materialize/pull/18719) or [materialize#17727](https://github.com/MaterializeInc/materialize/pull/17727) for examples.
+* [materialize#18578](https://github.com/MaterializeInc/materialize/pull/18578)
+  * Fixed previous migrations that failed to remove old types from the Stash.
+* `_incident-47`
+  * Scaling issue, the SQL statement we were running on CockroachDB was >32MiB which was the allowed size.
+
+
+# Explanation
+[explanation]: #explanation
+
+### What is the "Stash"?
+
+The [Stash](https://github.com/MaterializeInc/materialize/blob/23d7cad1211b4d1d1de3a26652937f9f2c4df61c/src/stash/src/lib.rs#L107-L124) 
+is a time-varying key-value store, i.e. it stores colletions of `(key, value, timestamp, diff)`, 
+which we use to persist metadata for a Materialize deployment. For example, we use the Stash to
+persist metadata for all `MATERIALIZED VIEWS` that a user has created. This way when restarting an 
+environment, we know what Materialized Views to recreate on Compute. Concretely we use CockroachDB
+and all keys and values are serialized as JSON for human readability.
+
+### Part 1: Snapshotting Types
+
+What makes Stash migrations hard to write and hard to reason about, is that you need to define a
+type that supports deserializing the current Stash and serializing the new Stash. For example, if
+you want to add a field to a struct you need to wrap the new field in an `Option<...>`. This way
+your struct can deserialize when the field doesn't exist, yet provide the Stash with the new data.
+There currently is not a way to do a migration like, converting a type from a `u64` to an `enum`.
+
+The proposal to fix this is to "snapshot" our types, this would allow us to, in a type safe way,
+represent the current data we're reading from, and the new types we want to migrate to. Then
+we'd be able to do arbitrarily complex migrations from one type to another.
+
+Concretely what this means is having types like:
+
+```
+struct RoleValuesV5 {
+    create_db: bool,
+}
+
+struct RoleValuesV6 {
+    // A bit flag of permissions this user is allowed.
+    perms: u64,
+}
+```
+
+### Part 2: Structuring Migrations
+
+Today migrations are a [list of closures](https://github.com/MaterializeInc/materialize/blob/23d7cad1211b4d1d1de3a26652937f9f2c4df61c/src/adapter/src/catalog/storage.rs#L78), and we index
```

Comment:
Where does `migrate.rs` fit into this proposal?

https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/migrate.rs

Reply:
AFAICT `migrate.rs` more so handles migrations of the Catalog, in other words, we've read in all of the data from the Stash and now treat everything as SQL objects, and operate on them as so. Whereas Stash migrations operate on "primitive" data, e.g. Strings, ints, and bools, and have no context to how they might relate to SQL objects.

IMO both code paths can and should exist separately, because they operate on the data at different layers.